File: solutions/Leetcode_827/leetcode_827.py
from collections import defaultdict, deque
from typing import List


class Solution1:
    directions = ((1, 0), (0, 1), (-1, 0), (0, -1))

    def largest_island(self, grid: List[List[int]]) -> int:
        self._grid = grid
        self._num_rows = len(grid)
        self._num_cols = len(grid[0])
        index = 2
        zeros = []
        index_area = {0: 0, 1: 0}
        for i in range(self._num_rows):
            for j in range(self._num_cols):
                if grid[i][j] == 1:
                    index_area[index] = self._dfs(i, j, index)
                    index += 1
                elif grid[i][j] == 0:
                    zeros.append((i, j))
        largest_area = max(index_area.values())
        for x, y in zeros:
            indices = set([grid[x+dx][y+dy] for dx, dy in Solution1.directions if self._check(x+dx, y+dy)])
            largest_area = max(largest_area, 1+sum([index_area[i] for i in indices]))
        return largest_area

    def _dfs(self, x, y, index):
        if self._check(x, y) and self._grid[x][y] == 1:
            self._grid[x][y] = index
            area = 1
            for dx, dy in Solution1.directions:
                area += self._dfs(x+dx, y+dy, index)
        else:
            area = 0
        return area

    def _check(self, x, y):
        if 0 <= x < self._num_rows and 0 <= y < self._num_cols:
            return True
        return False


class SolutionMINE:
    def largest_island(self, grid: List[List[int]]):
        zeros_adjacency = defaultdict(int)
        num_rows = len(grid)
        num_cols = len(grid[0])

        ones = []
        visited = {}
        for i in range(num_rows):
            for j in range(num_cols):
                if grid[i][j] == 1:
                    ones.append((i, j))

        if len(ones) == num_rows * num_cols:
            return num_rows * num_cols
        elif len(ones) == 0:
            return 1

        directions = ((1, 0), (0, 1), (-1, 0), (0, -1))
        while ones:
            x, y = ones.pop()
            if visited.get((x, y), False):
                continue
            q = deque([(x, y)])
            visited[(x, y)] = True
            area = 0
            neighbour_zeros = []
            while q:
                x, y = q.pop()
                area += 1
                for dx, dy in directions:
                    xx = x + dx
                    yy = y + dy
                    if 0 <= xx < num_rows and 0 <= yy < num_cols:
                        if grid[xx][yy] == 1 and not visited.get((xx, yy), False):
                            visited[(xx, yy)] = True
                            q.appendleft((xx, yy))
                        elif grid[xx][yy] == 0 and (xx, yy) not in neighbour_zeros:
                            neighbour_zeros.append((xx, yy))
            for neighbour in neighbour_zeros:
                zeros_adjacency[neighbour] += area

        return max(zeros_adjacency.values()) + 1

File: solutions/Leetcode_827/leetcode_827.cpp
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

class Solution1
{
public:
    int largestIsland(std::vector<std::vector<int>> &grid)
    {
        int num_rows = grid.size(), num_cols = grid[0].size(), index = 2;
        std::unordered_map<int, int> index_area;
        std::vector<std::pair<int, int>> zeros;
        index_area[0] = 0;
        index_area[1] = 0;
        for (int x = 0; x < num_rows; x++)
        {
            for (int y = 0; y < num_cols; y++)
            {
                if (grid[x][y] == 1)
                {
                    index_area[index] = dfs(grid, x, y, index);
                    index++;
                }
                else if (grid[x][y] == 0)
                {
                    zeros.push_back({x, y});
                }
            }
        }
        int largest_area = 0;
        for (const std::pair<const int, int> &x : index_area)
            largest_area = std::max(largest_area, x.second);
        for (const std::pair<int, int> &coor : zeros)
        {
            std::unordered_set<int> seen_index = {};
            int area = 1;
            for (int i = 0; i < 4; i++)
            {
                if (check(grid, coor.first + directions[i], coor.second + directions[i + 1]))
                {
                    index = grid[coor.first + directions[i]][coor.second + directions[i + 1]];
                    if (index > 1 && seen_index.find(index) == seen_index.end())
                    {
                        seen_index.insert(index);
                        area += index_area[index];
                    }
                }
            }
            largest_area = std::max(largest_area, area);
        }
        return largest_area;
    }

private:
    int directions[5] = {1, 0, -1, 0, 1};

    int dfs(std::vector<std::vector<int>> &grid, int x, int y, int index)
    {
        int area = 0;
        if (check(grid, x, y) && grid[x][y] == 1)
        {
            area++;
            grid[x][y] = index;
            for (int i = 0; i < 4; i++)
            {
                area += dfs(grid, x + directions[i], y + directions[i + 1], index);
            }
        }
        return area;
    }

    bool check(std::vector<std::vector<int>> &grid, int x, int y)
    {
        if (0 <= x && x < grid.size() && 0 <= y && y < grid[0].size())
            return true;
        return false;
    }
};
